zerver/views/streams.py

```diff
@@ -580,6 +582,147 @@ def you_were_just_subscribed_message(
+        color_map[channel.name] = channel.color
+
+    stream_dict_copy: StreamDict = {}
+    stream_dict_copy["name"] = channel.name.strip()
+
+    # We don't allow newline characters in stream descriptions.
+    if channel.description is not None:
+        stream_dict_copy["description"] = channel.description.replace("\n", " ")
+
+    stream_dict_copy["invite_only"] = invite_only
+    stream_dict_copy["is_web_public"] = is_web_public
+    stream_dict_copy["history_public_to_subscribers"] = history_public_to_subscribers
+    stream_dict_copy["message_retention_days"] = parse_message_retention_days(
+        message_retention_days, Stream.MESSAGE_RETENTION_SPECIAL_VALUES_MAP
+    )
+    stream_dict_copy["can_add_subscribers_group"] = group_settings_map["can_add_subscribers_group"]
+    stream_dict_copy["can_administer_channel_group"] = group_settings_map[
+        "can_administer_channel_group"
+    ]
+    stream_dict_copy["can_send_message_group"] = group_settings_map["can_send_message_group"]
+    stream_dict_copy["can_remove_subscribers_group"] = group_settings_map[
+        "can_remove_subscribers_group"
+    ]
+    stream_dicts.append(stream_dict_copy)
+
+    # Validation of the streams arguments, including enforcement of
+    # can_create_streams policy and check_stream_name policy is inside
+    # list_to_streams.
+    _, created_streams = list_to_streams(
+        stream_dicts,
+        user_profile,
+        autocreate=True,
+        is_default_stream=is_default_stream,
+        setting_groups_dict=setting_groups_dict,
+    )
+
+    is_subscribing_other_users = False
+    if len(subscribers) > 0 and not all(user_id == user_profile.id for user_id in subscribers):
+        is_subscribing_other_users = True
+
+    streams = created_streams
+
+    if is_default_stream:
+        do_add_default_stream(created_streams[0])
+
+    if is_subscribing_other_users:
+        new_subscribers = bulk_principals_to_user_profiles(subscribers, user_profile)
+    else:
+        new_subscribers = {user_profile}
+
+    (subscribed, _) = bulk_add_subscriptions(
+        realm, streams, new_subscribers, acting_user=user_profile, color_map=color_map
+    )
+    result: dict[str, Any] = dict(subscribed=defaultdict(list))
+    id_to_user_profile: dict[str, UserProfile] = {}
+    for sub_info in subscribed:
+        subscriber = sub_info.user
+        stream = sub_info.stream
+        user_id = str(subscriber.id)
+        result["subscribed"][user_id].append(stream.name)
+        id_to_user_profile[user_id] = subscriber
+    result["subscribed"] = dict(result["subscribed"])
+    send_messages_for_new_subscribers(
+        user_profile=user_profile,
+        subscribers=new_subscribers,
+        new_subscriptions=result["subscribed"],
+        id_to_user_profile=id_to_user_profile,
+        created_streams=created_streams,
+        announce=announce,
+    )
+    return json_success(
+        request,
+        data={"id": created_streams[0].id},
+    )
```

Comment:
We need to deduplicate this from the other code path more, maybe using some new helpers. Otherwise, I'm sure we'll eventually introduce bugs by updating one but not the other.

Reply:
It looks like you didn't do this. I added a commit extracting the main block of duplicated code as a function from `add_subscriptions_backend`.